akd/src/storage/transaction.rs

```diff
@@ -13,25 +13,23 @@ use crate::storage::types::ValueState;
 use crate::storage::types::ValueStateRetrievalFlag;
 use crate::storage::Storable;
 
+use dashmap::DashMap;
 use log::{debug, error, info, trace, warn};
 use std::collections::HashMap;
-use tokio::sync::RwLock;
-
-#[derive(Default, Clone)]
-struct TransactionState {
-    mods: HashMap<Vec<u8>, DbRecord>,
-    active: bool,
-}
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Arc;
 
 /// Represents an in-memory transaction, keeping a mutable state
 /// of the changes. When you "commit" this transaction, you return the
 /// collection of values which need to be written to the storage layer
 /// including all mutations. Rollback simply empties the transaction state.
+#[derive(Clone)]
 pub struct Transaction {
-    state: RwLock<TransactionState>,
+    mods: Arc<DashMap<Vec<u8>, DbRecord>>,
```

Comment:
Actually looking at it, nothing in this impl should be async anymore. Which is awesome! It'll save a lot of polling all around the place, it's now all sync work

Reply:
That's a good point, I've pushed the change! I've also made the calling functions sync if they no longer need to be async.